ZipCodeInputコンポーネントの作成

WATA-Haru opened the pull request:
## 実装内容
郵便番号入力フォームを作成しました。
https://github.com/user-attachments/assets/45337b40-2c9c-4b5c-bff3-7936ad2e6ae2

### 機能
1. 前半は3桁未満、後半は4桁未満の時エラーを表示する
2. 全角数字が入力された場合、自動で半角に変換する
3. 前半3桁が入力されたら、自動で次の後半4桁フォームに入力を移動する

### ディレクトリ構造
```
.
├── container
│   ├── composables
│   │   ├── useInputFirst.test.ts  // composablesの単体テスト(watchのみ)
│   │   ├── useInputFirst.ts
│   │   ├── useInputSecond.test.ts // composablesの単体テスト(watchのみ)
│   │   └── useInputSecond.ts
│   ├── index.test.ts
│   └── index.vue  // VueTestUtilsのContainerのテスト
└── presentational
    ├── index.test.ts // VueTestUtilsのPresentationalのテスト
    └── index.vue 
```

### review観点
1. **VueTestUtilsの書き方でよりよくかける方法があるか**
- 初めてなので推奨されていないパターンを使っているなど改善点が多いと思います。改善できる点を教えていただきたいです。

2. **テストの分け方が適切かどうか**
以下の3基準で分けたのですが適切かを見ていただきたいです。
`presentational`のテスト   
- `props`によるレンダリングが適切に行われこと
- 意図した条件下でカスタム`event`が発行されること

`container`のテスト
- テストでDOMをモックする必要があるもの
  - (ex. inputイベントで入力中の状態をモックするために`new InputEvent('input')`して`inputEvent.isComposing`を`Object.property`でセットするなど)

`composables`のテスト
- DOMをモックする必要がないテスト

3. **コンポーネントの分け方や共通部分の処理**
- 今回実装の重複がかなりあるように思います。共通化やもっと簡潔に欠ける点があればご指摘いただきたいです！

### (補足) input要素で取り扱うイベントについて

- 全角を半角に変換すること
- 文字列の長さが短いか、半角数字以外がある場合にエラーを設定する
ことがこのコンポーネントの責務で、下記のイベントでエラー検知や変換が行われる。

1. `input`イベント
2. `compositionEnd`イベント
3. `blur`イベント
4. エラー状態の`watch`

#### 1. `input`イベント
半角英数字を打つときに発火します。   
入力中は変換が起こらないようにしたいので[`isComposint`](https://developer.mozilla.org/en-US/docs/Web/API/InputEvent/isComposing)プロパティを使って入力中かを判定する。

#### 2. `compositionend`イベント
半角英数字以外の入力に使用します。   
IMEの入力確定時に[`Compositionend`](https://developer.mozilla.org/en-US/docs/Web/API/Element/compositionend_event)は発火する。

#### 3. `blur`イベント
`blur`の際に文字列の長さがminLengthを超えるかどうかを判定し、長さがminLengthよりも小さかった場合はエラーを設定する。

#### 4. エラー状態のwatch`
イベントではないが、inputのrefが更新された場合にエラーだったらエラー状態が`watch`で検知される。`watch`は全角半角数字が入力されていることと、minLengthの長さを検知する。なお、エラーのセットは`blur`が担当するため、文字数の長さのエラーは解除のみ行う。

#### 補足: 文字列の返還タイミングに`@change`や`@blur`を判定に使わない理由
`@blur`, `@keydown.enter`, `@change` で変換を検知した場合の問題
- 一度行った入力を消して再度同じ入力をするとイベントが発生しなくなる
- 例えば、入力欄に`１２３`などを入れた後にこれらを消し、再度`１２３`を入力すると入力前後で値に違いがないためイベントが発火しなくなる。

`@compositionend` で変換を検知した場合の問題
- 同じく、一度行った入力を消して再度同じ入力をするとイベントが発生しなくなる
- 半角数字の入力ではイベントが発火しなくなる

`@input`だけで判定する場合の問題
- 常にイベントが発行され、入力を連打するとminLengthを超えた長さの文字列が入力されてしまう(`isComposing`を設定すれば防げる)
- IMEの入力確定時にイベントが発火しなくなる

### (補足) Vitest In Source Testing
[In Source Testing](https://vitest.dev/guide/in-source)

composablesの中の関数にテストを書きました。
テストが実装の詳細を知りすぎな気がしたので書くか迷いましたが、   
エラー状態を切り替える関数であり、重要なもののエラー状態の切り替えを行う関数をexportしたくなかったのでinSourceTestingで書いた意図があります。

WATA-Haru committed:
[ADD] 全角数字を半角数字(文字列型)にして返す関数の作成

String(number)で変換すると'０１２'が'12'になる問題があったので自作した

WATA-Haru committed:
[ADD] InputFirstとInputSecondに共通するcomposablesの作成

WATA-Haru committed:
[ADD] 1つ目の入力欄を作成

WATA-Haru committed:
[ADD] 2つめの入力欄を追加

WATA-Haru committed:
[ADD] InputFirstとInputSecondをまとめるZipCodeInputを作成

WATA-Haru committed:
[ADD] Form(ZipCodeInputの上位の要素)を作成

WATA-Haru committed:
[ADD] App.vueにFormを追加

hanzzoo commented on A15/src/components/InputFirstHalf/container/index.vue:
```diff
@@ -0,0 +1,40 @@
+const { textRef, handleCompositionEnd, handleBlur } = useHalfInput(props.minLength, emitCallBack)
+</script>
+<template>
+  <InputFirstHalfPresentational
```
[ask]
propsの `:go-focus` が定義されてて渡さないとTypeErrorになるはずなんだけど、わたちゃんの環境ではTypeErrorででない？
CommitLintで `eslint`走らせるか、push前にチェックする癖つけた方がよいかも

WATA-Haru replied:
LanguageServerProtocolの問題っぽいので
vscodeで公式推奨の設定に変えて修正しました。

hanzzoo commented on A15/src/components/InputFirstHalf/container/index.vue:
```diff
@@ -0,0 +1,40 @@
+  } else if (eventName === 'input-shorter') {
+    emits('input-shorter')
+  }
+}
```
[imo]
`container` / `presentational` パターンにおいての `emits` の使い方がちょっと違うかな...

`emits`は実際のイベントハンドラーとコンポーネントのイベントをバインドさせるために使うものなので、
コールバックを `emits` で実装したことによって逆にイベント伝搬の流れがわかりづらくなっている気がするな...

hanzzoo commented on A15/src/components/InputFirstHalf/presentational/index.vue:
```diff
@@ -0,0 +1,81 @@
+const props = defineProps<Props>()
+const emits = defineEmits<Emits>()
+
+const isCompositionActive = ref(false)
```
[nits]
上手く `container/presentation`パターン出来ない理由として `presentation` に状態を持たせてしまうからだと思う。
`isCompositionActive` は `container`側で定義、操作して  `presentation`  は `props`を受け取るだけにした方が
データフローがシンプルになるし単体テストもしやすいよ

hanzzoo commented on A15/src/components/InputFirstHalf/presentational/index.vue:
```diff
@@ -0,0 +1,81 @@
+  emits('input-blur', target.value)
+}
+
+const secondInput = useTemplateRef('second-input')
```
[nits]
`useTemplateRef` もカスタムフックなので `container`側で定義した方がよき

hanzzoo commented on A15/src/components/InputFirstHalf/presentational/index.vue:
```diff
@@ -0,0 +1,81 @@
+  } else {
+    return
+  }
+})
```
[nits]
`watchEffect`は副作用なので、`presentation`に持たない方がよき

WATA-Haru commented:
ロジックとUIが分離された状態に書き直します！

WATA-Haru committed:
[DELETE] InputFirstHalf InputSecondHalf composables archieve すべて削除

コンポーネント設計を失敗したのですべて削除

WATA-Haru committed:
[ADD] 郵便番号全角数字変換と入力後フォーカス実装

WATA-Haru committed:
[FIX] IME以外の入力でもgoFocusを有効化

WATA-Haru committed:
[ADD] 文字数と数字以外の文字のバリデーション追加

WATA-Haru committed:
[REFACTOR] ref, watch等をhooksに切り離す

WATA-Haru committed:
[DELETE] archieve directory

WATA-Haru committed:
[FIX] テストケースのtypo修正

WATA-Haru committed:
[ADD] textRef*Watcherのdescriptionを追加

WATA-Haru committed:
[REFACTOR] errorStatusとgoFocusをInputFirst,InputSecondにひとまとめにするように変更

分離して書いていたがテストがしにくいためInputFirst,
Second内部でRefを変更するように修正

WATA-Haru committed:
[ADD] useInputFirst useInputSecondのプライベート関数へのテスト

vitest in source testingを使用ref: https://vitest.dev/guide/in-source

WATA-Haru committed:
[ADD] useInputFirstへの単体テスト追加

WATA-Haru committed:
[ADD] useInputSecondのテスト追加

WATA-Haru committed:
[FIX] isComposingの条件分岐を早期リターンに変更

WATA-Haru committed:
[FIX] reset-go-focusイベント名をlowerCamelCaseからkebab-caseに修正

WATA-Haru committed:
[FIX] testの際にEvent型の厳密な型チェックを行わないように設定

Jsdomの環境だとEventがInputEventやFocusEventではなくEventで解釈されるので厳密な型チェックを通らなくなってしまう

WATA-Haru committed:
[ADD] イベント発行できるかのテストを追加

WATA-Haru committed:
[FIX] 余分なasync削除とフォーマット

WATA-Haru committed:
[FIX] 条件分岐のisTestingのtrue, falseを逆にしていた部分を修正

WATA-Haru committed:
[FIX] isSomethingErrorの式でinvalidCharacterSecondを書いていなかったので修正

WATA-Haru committed:
[ADD] errorによる条件分岐と描画のテスト

WATA-Haru committed:
[ADD] HTMLイベントが絡むContainerComponentのテスト

WATA-Haru committed:
[DELETE] composablesのテストをwatchのみに

他のHTMLをモックするテストはvueTestUtilsに移動

hanzzoo commented on A15/src/components/ZipCodeInput/container/composables/useInputFirst.ts:
```diff
@@ -0,0 +1,197 @@
+      expect(goFocusRef.value).toBe(false)
+    })
+  })
+}
```
[nits]
composable関数の中にテストを書くのはやめた方がいいと思う。（可読性、保守性の観点で）
`useGoFocus` を別のファイルに切り出してそれと対になるテストファイルを用意するべき。

WATA-Haru replied:
```
composables/
  useInput/
    composables/goFocus.ts
```
下記コメントとかぶって恐縮ですが
このような形でcomposablesの中にcomposablesを入れる形で修正しました!
[b0b3b74](https://github.com/WATA-Haru/reviewMe/pull/20/commits/b0b3b74e0cd35bc4342d2875e8a0b0bf6d3fe1c6)

hanzzoo commented on A15/src/components/ZipCodeInput/container/composables/useInputFirst.ts:
```diff
@@ -0,0 +1,197 @@
+    unmarkInvalid,
+    markInvalid,
+  }
+}
```
[nits]
こちらも `useGoFocus` と同様で、別のファイルとテストに切り出そう。

WATA-Haru replied:
```
composables/
  useInput/
    composables/goFocus.ts
```

のような形で`composables`の中に`composables`を入れる形で修正しました!
[b0b3b74](https://github.com/WATA-Haru/reviewMe/pull/20/commits/b0b3b74e0cd35bc4342d2875e8a0b0bf6d3fe1c6)

hanzzoo commented on A15/src/components/ZipCodeInput/container/composables/useInputFirst.ts:
```diff
@@ -0,0 +1,197 @@
+  }
+}
+
+if (import.meta.vitest) {
```
[nits]
`composable`の中で テスト書いてる都合上だとは思うけど 通常のテスト書く時、わざわざわ環境変数を渡す必要はないです。

WATA-Haru replied:
ありがとうございます。InSourceTestingやめました！

[DELETE] vitest InSourceTesting用の設定を削除
[9c6914f](https://github.com/WATA-Haru/reviewMe/pull/20/commits/9c6914f53caf40924a7f2adb2fd83cfea102e30a)

こちらで設定を削除し、以下でcomposablesのテストを全て書きなおしました。
[FIX] useInputFirst関数を再度作成(副作用切り出し) 
[b0b3b74](https://github.com/WATA-Haru/reviewMe/pull/20/commits/b0b3b74e0cd35bc4342d2875e8a0b0bf6d3fe1c6)

[FIX] useInputSecond関数を再度作成(副作用切り出し)
[30d8bde](https://github.com/WATA-Haru/reviewMe/pull/20/commits/30d8bde87624a005f9eed0612f45c954717e82d1)

hanzzoo commented on A15/src/components/ZipCodeInput/container/composables/useInputFirst.ts:
```diff
@@ -0,0 +1,197 @@
+    handleInputFirst,
+    handleCompositionEndFirst,
+    handleBlurFirst,
+    textRefFirstWatcher,
```
[imo]
`watch` を行う機能 を composable関数の戻り値にしている理由って何かある？

自分の解釈では、`watch` は、 特定の状態 / プロパティの変更を監視するので、
コンポーネントのライフサイクルと密になるという特性があると思っていて
特に明示的に戻り値にしなくても良いと思っている。

WATA-Haru replied:
確かにおっしゃる通りです。必要なかったので削除しました

Container側での呼び出し削除
[e0f96da](https://github.com/WATA-Haru/reviewMe/pull/20/commits/e0f96daba82e3aa5e1e3f119fb2994fa1f642457)

`useInputFirst/index.ts`の修正
[b0b3b74](https://github.com/WATA-Haru/reviewMe/pull/20/commits/b0b3b74e0cd35bc4342d2875e8a0b0bf6d3fe1c6)

`useInputSecond/index.ts`の修正
[30d8bde](https://github.com/WATA-Haru/reviewMe/pull/20/commits/30d8bde87624a005f9eed0612f45c954717e82d1)

hanzzoo commented on A15/src/components/ZipCodeInput/container/composables/useInputFirst.test.ts:
```diff
@@ -0,0 +1,32 @@
+import { nextTick } from 'vue'
+
+describe('textRefFirstWatcher', () => {
+  test('入力値が全角・半角数字以外の場合、markInvalid が呼ばれて errorRefFirst.invalid が true になること', async () => {
```
[imo]
`markInvalidが呼ばれて~` テストケースの部分が `useInputFirst` の実装のインターフェースに現れていないので、
テストケースが composable内部の実装詳細を知りすぎていると思ったけど、どうだろう？

実装していない人がこのテストケースを見た時に、
確実にcomposable関数の内部処理を追いにいくと思うな...

hanzzoo replied:
下のテストケースの `unmarkInvalid` も同様

WATA-Haru replied:
確かにmarkInvalidが外から見えない状態なのにテストケースに書いているのはよくないですね。


下記コメントとかぶって恐縮ですが下記で対応しました
こちらで設定を削除し、以下でcomposablesのテストを全て書きなおしました。
[FIX] useInputFirst関数を再度作成(副作用切り出し) 
[b0b3b74](https://github.com/WATA-Haru/reviewMe/pull/20/commits/b0b3b74e0cd35bc4342d2875e8a0b0bf6d3fe1c6)

[FIX] useInputSecond関数を再度作成(副作用切り出し)
[30d8bde](https://github.com/WATA-Haru/reviewMe/pull/20/commits/30d8bde87624a005f9eed0612f45c954717e82d1)

hanzzoo commented on A15/src/components/ZipCodeInput/container/composables/useInputFirst.test.ts:
```diff
@@ -0,0 +1,32 @@
+import { useInputFirst } from './useInputFirst'
+import { nextTick } from 'vue'
+
+describe('textRefFirstWatcher', () => {
```
[imo]
describe は `useInputFirst` であるのが正しいかな...

WATA-Haru replied:
ありがとうございます。間違っていますね...

かぶりますが、composablesのテストを全て書きなおし、その際に一緒に修正しました。

[FIX] useInputFirst関数を再度作成(副作用切り出し)
[b0b3b74](https://github.com/WATA-Haru/reviewMe/pull/20/commits/b0b3b74e0cd35bc4342d2875e8a0b0bf6d3fe1c6)

[FIX] useInputSecond関数を再度作成(副作用切り出し)
[30d8bde](https://github.com/WATA-Haru/reviewMe/pull/20/commits/30d8bde87624a005f9eed0612f45c954717e82d1)

hanzzoo commented on A15/src/components/ZipCodeInput/presentational/index.vue:
```diff
@@ -0,0 +1,141 @@
+  isInvalidCharacterUsedSecond: boolean
+  isShorterThanMinLengthSecond: boolean
+}
+interface Emits {
```
[nits]
 `Emits` のインターフェースを型定義するときは
以下の方みたいな書き方が慣習だと思う。
```ts
interface Emits {
  (eventName: 'input-first', event: Event): void
}
```

` 'input-first'` は Vueのイベントを発行させるための 識別子であって `event` オブジェクトではないので

WATA-Haru replied:
修正しました！
[4a3ad83](https://github.com/WATA-Haru/reviewMe/pull/20/commits/4a3ad83aa30f08e79f2c3467139ef7d966a3483b)

hanzzoo commented on A15/src/components/ZipCodeInput/presentational/index.vue:
```diff
@@ -0,0 +1,141 @@
+    return
+  }
+  emits('input-first', event as Readonly<InputEvent>)
+}
```
[q]
どういう理由でこうしてる...？

コンポーネント側で `@input="(event: Event) => handleInputFirst(event)"` と渡しているなら、
素直に以下で良いのではと思った。

```ts
const handleInputFirst = (event: Event) => {
  emits('input-first', event)
}
```

hanzzoo replied:
もし インスタンスタイプをできるだけ縛りたいとしても、せめてこうとか...

```ts
interface Emits { (eventName: 'input-first', event: InputEvent): void }

const handleInputFirst = (event: Event) => {
  if (!(event instanceof InputEvent)) {
    return
  }
  emits('input-first', event)
}
```

WATA-Haru replied:
Emitが渡されたEventオブジェクトに変更可能性があると誤認していました([Readonlyでした](https://developer.mozilla.org/ja/docs/Web/API/Event))


変更可能性はないので下記で修正しました！
[4a3ad83](https://github.com/WATA-Haru/reviewMe/pull/20/commits/4a3ad83aa30f08e79f2c3467139ef7d966a3483b)

hanzzoo commented on A15/src/components/ZipCodeInput/presentational/index.test.ts:
```diff
@@ -0,0 +1,190 @@
+    isShorterThanMinLengthSecond: false,
+  }
+
+  test('input-first: first input に値がセットされたとき、イベントが発行されること', async () => {
+    const wrapper = mount(ZipCodeInputPresentational, { props: defaultProps })
+    const inputFirst = wrapper.find('.zipcode-input-wrapper__input-first')
+    await inputFirst.setValue('123')
+    expect(wrapper.emitted()).toHaveProperty('input-first')
+  })
```
[nits]
input発火時にイベントが発行されるかだけテストすればOKです。
input要素に `data-test-id` みたいな形でIDを振って、utils testの機能を使えます ( `trigger` )


```ts

  test('inputイベント実行時にemitが1回発行されること', () => {
    const wrapper = mount(ZipCodeInputPresentational, { props: defaultProps })
    const inputFirst = wrapper.find('[data-test-id="input-first"]')
    inputFirst.trigger('input')
    expect(wrapper.emitted('input')?.length).toBe(1)
  })

```

hanzzoo replied:
下の `reset-go-focus-status` までの テストケースも同様

WATA-Haru replied:
存じ上げませんでした。教えていただきありがとうございます。
[findByTestIDプラグイン](https://test-utils.vuejs.org/guide/extending-vtu/plugins)を導入してテストしてみました！

プラグイン導入
[0cc7975](https://github.com/WATA-Haru/reviewMe/pull/20/commits/0cc797547a16084a4981730866d394f7284a78a2)

`Presentational`のテストを`findByTestId`を使用したテストに変更
[4a3ad83](https://github.com/WATA-Haru/reviewMe/pull/20/commits/4a3ad83aa30f08e79f2c3467139ef7d966a3483b)

hanzzoo commented on A15/src/components/ZipCodeInput/container/composables/useInputSecond.ts:
```diff
@@ -0,0 +1,143 @@
+  }
+}
+
+if (import.meta.vitest) {
```
[nits]
`useInputFirst` と同様に 関数とテストファイルを分けよう

WATA-Haru replied:
`composables`の中に`composables`を入れる形で修正しました！
[30d8bde](https://github.com/WATA-Haru/reviewMe/pull/20/commits/30d8bde87624a005f9eed0612f45c954717e82d1)

hanzzoo commented on A15/src/components/ZipCodeInput/container/index.test.ts:
```diff
@@ -0,0 +1,357 @@
+const inputFirstClassName = 'zipcode-input-wrapper__input-first'
+const inputSecondClassName = 'zipcode-input-wrapper__input-second'
+
+describe('入力欄1: inputの入力確定時に文字列が変換されることのテスト', () => {
```
[nits]
`container` 側で DOMのテストをしてしまうと `container / presentational ` パターンが崩壊してしまうと思う。
`composable` で切り出している場合は `container` の テストはなくなるはず。

状態・(変換などの) ロジック処理により値を返すことをテストする =  `compsable` / `container` 
DOMのイベントの振る舞い、propsの描画をテストする = `presentation`

詳細にテストを書けばよいってわけでもなく、
テストコードも負債になっていきます。

WATA-Haru replied:
ありがとうございます。`composables`で持たせる方がよいですね...削除しました！

`Container`側のテストを削除
[8a2da62](https://github.com/WATA-Haru/reviewMe/pull/20/commits/8a2da6240bd709f538e0a0392f47ba204545d085)

`useInputFirst`
側にテストを記載
[b0b3b74](https://github.com/WATA-Haru/reviewMe/pull/20/commits/b0b3b74e0cd35bc4342d2875e8a0b0bf6d3fe1c6)

`useInputSecond`
側にテストを記載
[30d8bde](https://github.com/WATA-Haru/reviewMe/pull/20/commits/30d8bde87624a005f9eed0612f45c954717e82d1)

hanzzoo reviewed (commented):
とりあえず見切れる範囲でコメントしたけど、
big PRは避けていこう！

WATA-Haru commented on A15/src/utils/makeFullWidthNumToHalfWidthNum/index.ts:
```diff
@@ -0,0 +1,20 @@
+
+/**
+ * @param value string
+ * @return string
```
[memo]
特殊な引数・型じゃなかったらいらないかも(`ts`で担保できるから)

WATA-Haru commented on A15/src/utils/makeFullWidthNumToHalfWidthNum/index.ts:
```diff
@@ -0,0 +1,20 @@
+/**
+ * @param value string
+ * @return string
+ * @description - 全角の数字を半角の数字にして返す関数、String(number)で直接変換すると先頭の０が消えるため作成(ex. '012'が'12'になってしまう)
```
[memo]
`@example`というタグがある

WATA-Haru commented on A15/src/components/ZipCodeInput/container/composables/useInputSecond.ts:
```diff
@@ -0,0 +1,143 @@
+}
+
+if (import.meta.vitest) {
+  const { describe, test, expect } = import.meta.vitest
```
[memo] InSourceTesting使うくらいだったらcomposablesの中にcomposables入れたディレクトリ構成にしてよさそう
```
composables/
  useInput/
    composables/goFocus.ts
```

WATA-Haru committed:
[DELETE] vitest InSourceTesting用の設定を削除

WATA-Haru committed:
[ADD] findByTestIdプラグインの導入

WATA-Haru committed:
[FIX] 不必要な型チェックの削除

WATA-Haru committed:
[FIX] findByTestIdプラグイン導入とEvent呼び出し時のテスト簡略化

WATA-Haru committed:
[DELETE] containerのテスト削除(composablesのテストをすべきなのでcontainerのテスト削除)

WATA-Haru committed:
[FIX] useInputFirst関数を再度作成(副作用切り出し)

- InSourceTestingを削除
- composablesの中にcomposablesを作成し副作用切り出し
- containerではなくcomposables側にテスト書き直し

WATA-Haru committed:
[FIX] useInputSecond関数を再度作成(副作用切り出し)

- InSourceTestingを削除
- composablesの中にcomposablesを作成し副作用切り出し
- containerではなくcomposables側にテスト書き直し

WATA-Haru committed:
[DELETE] watcherの削除

hanzzoo reviewed (approved):
![LGTM](https://image.lgtmoon.dev/315367)
